File: daily/2019-12-29/6 ZigZag Conversion.py
class Solution(object):
    def convert(self, s, numRows):
        if numRows == 1 or numRows >= len(s):
            return s
    
        zigzag = ['' for x in range(numRows)]
        row, step = 0, 1
        for char in s:
            zigzag[row] += char
            if row == 0:
                step = 1
            elif row == numRows - 1:
                step = -1
            row += step
        return ''.join(zigzag)

File: Leetcode First 200 JS solution/6. Zigzag Conversion.js
var convert = function(s, numRows) {
    
    if (numRows == 1 || s.length == 1) return s;

    let stack = [];
    for (let i = 0; i < numRows; i++) stack.push('');
    let step = 1, row = 0;
    for (let char of s){
        stack[row] += char;
        if (row == 0) {
            step = 1;
        } else if (row == numRows - 1) {
            step = -1
        }
    }
    return stack.join('');
}

// o(n) time
// o(n) space